lib_i2s/src/i2s_frame_slave_impl.h

```diff
@@ -74,85 +85,216 @@ static void i2s_frame_slave0(client i2s_frame_callback_if i2s_i,
+            {
+                p_dout[i] <: bitrev(out_samps[idx]);
+            }
         }
+        else
+        {
+            // Start outputting evens (0,2,4..) data at correct point relative to the clock
+#pragma loop unroll
+            for (size_t i = 0, idx = 0; i < num_out; i++, idx += 2)
+            {
+                partout_timed(p_dout[i], num_data_bits, bitrev(out_samps[idx] << data_bit_offset), initial_out_port_time);
+            }
+
+            // XC doesn't have syntax for setting a timed input without waiting for the input 
+            asm volatile("setpt res[%0], %1"
+                        :
+                        :"r"(p_lrclk),"r"(initial_in_port_time));
+            set_port_shift_count(p_lrclk, num_data_bits);
+
+            for (size_t i = 0; i < num_in; i++) 
+            {
+                asm volatile("setpt res[%0], %1"
+                            :
+                            :"r"(p_din[i]),"r"(initial_in_port_time));
+                set_port_shift_count(p_din[i], num_data_bits);
+            }
 
-        //And pre-load the odds (1,3,5..) to follow immediately afterwards
-        for (size_t i=0, idx=1; i<num_out; i++, idx+=2){
-            p_dout[i] <: bitrev(out_samps[idx]);
+            //And pre-load the odds (1,3,5..) to follow immediately afterwards
+            for (size_t i = 0, idx = 1; i < num_out; i++, idx += 2)
+            {
+                partout(p_dout[i], num_data_bits, bitrev(out_samps[idx] << data_bit_offset));
+            }
         }
 
+
         //Main loop
-        while (!syncerror && (restart == I2S_NO_RESTART)) {
+        while (!syncerror && (restart == I2S_NO_RESTART)) 
+        {
             restart = i2s_i.restart_check();
 
-
-            if (num_out && (restart == I2S_NO_RESTART)){
+            if (num_out && (restart == I2S_NO_RESTART))
+            {
                 i2s_i.send(num_out << 1, out_samps);
 
-                //Output i2s evens (0,2,4..)
+                // Output i2s evens (0,2,4..)
+                if (num_data_bits == 32)
+                {
 #pragma loop unroll
-                for (size_t i=0, idx=0; i<num_out; i++, idx+=2){
-                    p_dout[i] <: bitrev(out_samps[idx]);
+                    for (size_t i = 0, idx = 0; i < num_out; i++, idx += 2)
+                    {
+                        p_dout[i] <: bitrev(out_samps[idx]);
+                    }
+                }
+                else
+                {
+#pragma loop unroll
+                    for (size_t i = 0, idx = 0; i < num_out; i++, idx += 2)
+                    {
+                        partout(p_dout[i], num_data_bits, bitrev(out_samps[idx] << data_bit_offset));
+                    }
                 }
             }
                 
-            //Read lrclk value
-            p_lrclk :> lrval;
+            // Read lrclk value
+            if (num_data_bits == 32)
+            {
+                p_lrclk :> lrval;
```

Comment:
I'm surprised to see no error checking of the read LRCLK value in the original code. One might expect to re-sync to LRclock if the value is not as-expected.

EDIT: ignore this - spotted it further down